File: Javascript/2019 카카오 겨울 인턴/n04.js
function solution(k, room_number) {
    let answer = [];
    let link = new Map();
    let tmp;
    let tmp_list;
    room_number.forEach(e => {
        if (!link.has(e)) {
            link.set(e, e+1);
            answer.push(e);
        } else {
            tmp = e;
            tmp_list = [];
            while (link.has(tmp)) {
                tmp_list.push(tmp);
                tmp = link.get(tmp);
            }
            link.set(tmp, tmp + 1);
            tmp_list.forEach(t => {
                link.set(t, tmp + 1);
            });
            answer.push(tmp);
        }
        console.log(link);
    });
    return answer;
}

// console.log(solution( // result [1,3,4,2,5,6]
//     10, 
//     [1,3,4,1,3,1]
// ));
console.log(solution( // result [1,2,3,4,5,6,7,8,9]
    100, 
    [1,1,1,1,5,5,5,5,3]
));

File: Python/2019 카카오 겨울 인턴/n04.py
def solution(k, room_number):
    answer = []
    room_dic = dict()
    for rm in room_number:
        if rm not in room_dic.keys():
            room_dic[rm] = rm+1
            answer.append(rm)
        else:
            i = rm
            key_list = []
            while i in room_dic.keys():
                key_list.append(i)
                i = room_dic[i]
            room_dic[i] = i+1
            for key in key_list:
                room_dic[key] = i+1
            answer.append(i)
        print(room_dic)
    return answer

if __name__ == "__main__":
    ans = solution(10, [1,3,4,1,3,1])
    # ans = solution(100, (1,1,1,1,5,5,5,5,3))
    print(ans)